### PR TITLE
Add pseudo-streams.

### PR DESCRIPTION
**Pull request by sunfishcode**
This add a pseudo-stream type to the wasi-poll interface, and adds ways to obtain streams from command invocation and from files. In the future, it can support sockets too. With this, `command` takes streams for stdin/stdout, rather than filesystem descriptors.

Streams support reading and writing, as well as skipping, repeated-element writing, and splicing from one stream to another. And there are `subscribe-*` functions to produce pseudo-futures from pseudo-streams, allowing them to be polled.

This makes the polyfill somewhat more complex, but this is largely due to the polyfill being tied to the preview1 API.

This replaces the `seek` and `tell` functions, and implemented `fd_seek` and `fd_tell` in terms of the polyfill's own position.

Also, add a dedicated stderr API for writing to stderr in a way that tolerates strings that aren't necessarily expected to be newlines. And add a way to test whether stderr is a terminal.

### CODE REVIEW

**Comment by sunfishcode** [comment]
This now also implements polling, on both streams and clocks.

**Review by dicej — APPROVED** [review]
Looks good overall.  Please see comments inline.

Also, (and this may be out of scope for this PR) I'm wondering how much of the existing `cap-std-sync::sched` implementation we want to keep given the comments in windows.rs indicate that it is buggy and unmaintained.  Not that I'm volunteering to replace it :).  Just wondering if we want to try to reuse something like Tokio instead.

File: host/src/clocks.rs [hunk 1]
```diff
@@ -19,102 +18,56 @@ impl TryFrom<SystemTime> for wasi_clocks::Datetime {
 
 #[async_trait::async_trait]
 impl wasi_default_clocks::WasiDefaultClocks for WasiCtx {
-    async fn default_monotonic_clock(&mut self) -> anyhow::Result<wasi_clocks::MonotonicClock> {
-        Ok(self.clocks.default_monotonic)
+    async fn default_wall_clock(&mut self) -> anyhow::Result<wasi_clocks::WallClock> {
+        // Create a new handle to the default wall clock.
+        let new = self.clocks.default_wall_clock.dup();
+        Ok(self.table_mut().push(Box::new(new))?)
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply host/src/clocks.rs:24]: I am expecting users will cache and reuse these functions, if they're needed frequently. Creating a new handle each time frees us from having to figure out what to do if the user closes the one handle we allocated up front.

I've now added a patch adding documentation, and https://github.com/WebAssembly/wasi-clocks/pull/22 proposing adding it to the upstream wasi-clocks proposal.

File: src/lib.rs [hunk 2]
```diff
@@ -695,32 +749,45 @@ pub unsafe extern "C" fn fd_read(
...
-            Descriptor::File(f) => f,
-            Descriptor::Closed(_) | Descriptor::StdoutLog | Descriptor::StderrLog => {
-                return Err(ERRNO_BADF)
+        match state.get(fd)? {
+            Descriptor::Streams(streams) => {
+                let wasi_stream = streams.get_read_stream()?;
+
+                let read_len = unwrap_result(u32::try_from(len));
+                let wasi_stream = streams.get_read_stream()?;
+                //let (data, end) =
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply src/lib.rs:758]: Yep.

File: host/src/poll.rs [hunk 3]
```diff
@@ -1,18 +1,186 @@
...
+        Ok(bytes_written)
+    }
+
+    async fn splice_stream(
+        &mut self,
+        _src: WasiStream,
+        _dst: WasiStream,
+        _len: u64,
+    ) -> HostResult<(u64, bool), StreamError> {
+        // TODO: We can't get two streams at the same time because they both
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply host/src/poll.rs:144]: We don't need `splice` for the preview1 polyfill so I'm thinking about leaving this code as a `todo` for now, but if we do need this, that's a good idea for now to do it.

**Comment by sunfishcode** [comment]
> Also, (and this may be out of scope for this PR) I'm wondering how much of the existing `cap-std-sync::sched` implementation we want to keep given the comments in windows.rs indicate that it is buggy and unmaintained. Not that I'm volunteering to replace it :). Just wondering if we want to try to reuse something like Tokio instead.

I think we do want to do this, and I wouldn't object if someone volunteered to do it, but if not I think we're ok there for the time being. The higher priorities are building up the preview2 API and getting to feature parity with preview1.

File: host/src/clocks.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sunfishcode) [reply host/src/clocks.rs:24]: Also yes, we should have a way to close these handles. I've now posted https://github.com/bytecodealliance/preview2-prototyping/pull/39.

**Pull request merged**